### ISSUE TITLE
Whether to support the conversion torchscript mixing accuracy

### ISSUE THREAD

**Issue by wangxiaodong1021** [NVIDIA/apex#303]
Whether to support the conversion torchscript mixing accuracy?

`model.eval()

# An example input you would normally provide to your model's forward() method.
example = torch.rand(1, 3, 224, 224)

# Use torch.jit.trace to generate a torch.jit.ScriptModule via tracing.
traced_script_module = torch.jit.trace(model, example)
traced_script_module.save("model.pt")`

**Comment by mcarilli** [comment]
Support for jit tracing is still being tested.  Creating a traced module via `torch.jit.trace` within a script, then using that traced model within the same script, should work as long as you call
```
model, optimizer = amp.initialize(model, optimizer, opt_level="O1")
```
**before**
```
traced_script_module = torch.jit.trace(model, example)
```
Beware, if the original model has any data-dependent control flow, that control flow will be ignored by the traced_script_module (but that's a limitation of all traced modules, it's not specific to models touched by Amp).

This is still very experimental for us.  I also don't know what happens if you try to save the model.

**Comment by justusschock** [comment]
Do you have any experience with using custom ScriptModules instead of tracing? We cannot call `amp.initialize` on the model this way. So it probably won't work correctly?

**Comment by NotNANtoN** [comment]
Hello @mcarilli, any updates on this? I created a custom TorchScript LSTM, which returns the cell states of all the timesteps along with the hidden states because using the LSTMCell was too slow. I think I don't know enough about TorchScript and Apex to really understand your hint. As far as I understand you suggest to first apply amp to the model and then do the jit torchscript tracing. Unfortunately, this might not be possible with my current setup:
```
class LSTMCell(jit.ScriptModule):
    def __init__(self, input_size, hidden_size):
        super(LSTMCell, self).__init__()
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.weight_ih = Parameter(torch.randn(4 * hidden_size, input_size))
        self.weight_hh = Parameter(torch.randn(4 * hidden_size, hidden_size))
        self.bias_ih = Parameter(torch.randn(4 * hidden_size))
        self.bias_hh = Parameter(torch.randn(4 * hidden_size))

    @jit.script_method
    def forward(self, input, state):
        # type: (Tensor, Tuple[Tensor, Tensor]) -> Tuple[Tensor, Tuple[Tensor, Tensor]]
        hx, cx = state
        gates = (torch.mm(input, self.weight_ih.t()) + self.bias_ih +
                 torch.mm(hx, self.weight_hh.t()) + self.bias_hh)
        ingate, forgetgate, cellgate, outgate = gates.chunk(4, 1)

        ingate = torch.sigmoid(ingate)
        forgetgate = torch.sigmoid(forgetgate)
        cellgate = torch.tanh(cellgate)
        outgate = torch.sigmoid(outgate)

        cy = (forgetgate * cx) + (ingate * cellgate)
        hy = outgate * torch.tanh(cy)

        return hy, (hy, cy)


class LSTMLayerStates(jit.ScriptModule):
    def __init__(self, cell, *cell_args):
        super(LSTMLayerStates, self).__init__()
        self.cell = cell(*cell_args)

    @jit.script_method
    def forward(self, input, state):
        # type: (Tensor, Tuple[Tensor, Tensor]) -> Tuple[Tensor, Tensor, Tuple[Tensor, Tensor]]
        inputs = input.unbind(0)
        outputs = torch.jit.annotate(List[Tensor], [])
        states = torch.jit.annotate(List[Tensor], [])
        for i in range(len(inputs)):
            out, state = self.cell(inputs[i], state)
            outputs += [out]
            states += [state[1]]
        return torch.stack(outputs), torch.stack(states), state
```

Is there any way to apply amp to the LSTMLayerStates? I am thinking that maybe I could remove the @jit.script_method's and jit annotations within the classes and then use torch.jit.trace(model, example) as you said. But I fear that even if this works with amp it will make the TorchScript part much slower.

I would be grateful for any advice!

**Comment by mcarilli** [comment]
Script mode isn't supported at all by Apex, because this requires a native integration.  Our main task at this time is native support for mixed precision.  I have PRs up for [gradient scaling](https://github.com/pytorch/pytorch/pull/26512) and [eager-mode autocasting](https://github.com/pytorch/pytorch/pull/29552).  @jjsjann123 is working on [jit script support](https://github.com/pytorch/pytorch/issues/25387).

**Comment by xsacha** [comment]
Is there a way to remove APEX from the model so that I can use torch.jit.trace? Currently it is crashing for me.
